packages/orderbook/src/engine/matching-engine.ts

```diff
@@ -316,68 +333,144 @@ export class MatchingEngine {
 		// this.bookInSync = false
 	}
 
-	async settle(buyOrder: LimitOrder, sellOrder: LimitOrder) {
-		console.log('Settlement', buyOrder, sellOrder)
+	async settle(matchedOrders: LimitOrder[]) {
+		const settlementId = randomBytes(16).toString('hex')
+
+		const orderIds = matchedOrders.map((order) => order.id)
+		if (orderIds.some((id) => this.isOrderInPendingSettlement(id))) {
+			logger.debug('Some orders are already in pending settlement')
+			return
+		}
+
+		this.pendingSettlements.set(settlementId, { orders: orderIds, retryCount: 0 })
+
+		const request = {
+			orders: matchedOrders.map((order) => order.order),
+			signatures: matchedOrders.map((order) => order.signature),
+		}
 
 		try {
 			const settlementSimulate = await baseClient.readContract({
 				address: verifyingContract,
 				abi: clearingHouseABI,
 				functionName: 'canSettle',
-				args: [
-					{
-						orders: [buyOrder.order, sellOrder.order],
-						signatures: [buyOrder.signature, sellOrder.signature],
-					},
-				],
+				args: [request],
 			})
 
-			console.log({
-				success: settlementSimulate.success,
-				data: fromHex(settlementSimulate.data, 'string'),
-			})
-		} catch (e) {
-			console.dir(e, { depth: Number.POSITIVE_INFINITY })
+			if (settlementSimulate.success) {
+				const settlement = await walletClient.writeContract({
+					address: verifyingContract,
+					abi: clearingHouseABI,
+					functionName: 'settle',
+					args: [request],
+				})
+
+				const receipt = await baseClient.waitForTransactionReceipt({
+					hash: settlement,
+				})
+
+				if (receipt.status !== 'success') {
+					throw new Error('Settlement failed')
+				}
+
+				logger.info(`Settlement ${settlementId} successful. Transaction hash: ${settlement}`)
+
+				for (const order of matchedOrders) {
+					await this.updateOrderStatus(order.id, 'executed', settlement)
+				}
+
+				this.pendingSettlements.delete(settlementId)
+			} else {
+				throw new Error(fromHex(settlementSimulate.data, 'string'))
+			}
+		} catch (e: unknown) {
+			logger.error(
+				`Settlement ${settlementId} failed: ${e instanceof Error ? e.message : 'Unknown error'}`
+			)
+			const pendingSettlement = this.pendingSettlements.get(settlementId)!
+
+			if (pendingSettlement.retryCount < 1) {
+				pendingSettlement.retryCount++
+				this.pendingSettlements.set(settlementId, pendingSettlement)
+				setTimeout(() => this.settle(matchedOrders), 5000)
+			} else {
+				for (const orderId of pendingSettlement.orders) {
+					await this.markOrderAsFailed(orderId)
+				}
+				this.pendingSettlements.delete(settlementId)
+			}
 		}
 	}
 
-	async checkForPossibleSettles() {
-		// if (this.bookClean) return
+	private isOrderInPendingSettlement(orderId: string): boolean {
+		for (const settlement of this.pendingSettlements.values()) {
+			if (settlement.orders.includes(orderId)) {
```

Comment:
Perhaps we should store them as a map to avoid the `O(n)` lookup inside of the above `.some` which will be `O(n^2)`